packages/jamtools/features/modules/dashboards/keytar_and_foot_dashboard/multi_octave_supervisor.tsx

```diff
@@ -0,0 +1,163 @@
+                    return e.number !== event.number;
+                }),
+            };
+        }
+
+        setTimeout(() => {
+            if (this.states.enableDebugging.getState()) {
+                this.states.debugSavedInputEvent.setState(fullEvent);
+                this.states.debugMidiState.setState(this.midiState);
+            }
+        });
+    };
+
+    private createActions = () => ({
+        toggleDebugging: this.moduleAPI.createAction(`${this.kvPrefix}|toggleDebugging`, {}, async () => {
+            console.log('toggling debug mode', !this.states.enableDebugging.getState());
+            this.states.enableDebugging.setState(!this.states.enableDebugging.getState());
+        }),
+    });
+
+    public render: React.ElementType = () => {
+        const enableDebugging = this.states.enableDebugging.useState();
+        const debugSavedInputEvent = this.states.debugSavedInputEvent.useState();
+        const debugMidiState = this.states.debugMidiState.useState();
+
+        return (
+            <>
+                <this.renderDebugData
+                    enableDebugging={enableDebugging}
+                    debugSavedInputEvent={debugSavedInputEvent}
+                    debugMidiState={debugMidiState}
+                />
+            </>
+        );
+    };
+
+    private renderDebugData = ({debugSavedInputEvent, enableDebugging, debugMidiState}: {debugSavedInputEvent: MidiEventFull | null, enableDebugging: boolean, debugMidiState: MultiOctaveSupervisorMidiState}) => {
+        const [showDebugData, setShowDebugData] = React.useState(false);
+
+        return (
+            <>
+                <Button onClick={() => this.actions.toggleDebugging({})}>
+                    {enableDebugging ? 'Disable debugging' : 'Enable debugging'}
+                </Button>
+
+                <Button onClick={() => setShowDebugData(!showDebugData)}>
+                    {showDebugData ? 'Hide debug data' : 'Show debug data'}
+                </Button>
+
+                {showDebugData && (
+                    <>
+                        <p>Paged octave input:</p>
+                        <div>
+                            <this.macros.pagedOctaveInput.components.edit />
+                        </div>
+
+                        <p>Paged octave output:</p>
+                        <div>
+                            <this.macros.midiOutput.components.edit />
+                        </div>
+
+                        {debugSavedInputEvent && (
+                            <pre>
+                                {JSON.stringify(debugSavedInputEvent, null, 2)}
+                            </pre>
+                        )}
+
+                        {debugMidiState && (
+                            <pre>
+                                {JSON.stringify(debugMidiState, null, 2)}
+                            </pre>
+                        )}
+                    </>
+                )}
+            </>
+        );
+    };
```

Comment:
Note that all of this react component stuff with macros will be handled by the framework, so developers won't have this code in their implementations. This is all boilerplate to allow the user to configure the macros behind the scenes

Reply:
Would be cool if it was really easy to see your macros and states of a given dashboard/module/snack in a debug view automatically. Something that allows developers to iterate quickly and understand their data flow. They don't need to render any of it. See state changes in realtime, with the ability to dot notation show fields

Investing in a "selectors" pattern is probably a good idea. Some pattern to get certain data. Would be better if you could mostly do your use cases directly from the `StateSupervisor` methods, and not need another abstraction.

Maybe make selectors a first class thing that can be inspected as well. Selectors that are being called and what they are returning. Maybe not, that sounds expensive for otherwise quick operations

Maybe also have some sort of time travel thing if possible. The state supervisors just need to coordinate. You'll want to batch things somehow, if you want time travel to work well